scripts/data_utils.py

```diff
@@ -36,11 +38,226 @@
 SENTENCE_ENDINGS = [".", "!", "?"]
 WORDS_BREAKS = list(reversed([",", ";", ":", " ", "(", ")", "[", "]", "{", "}", "\t", "\n"]))
 
+HTML_TABLE_TAGS = {"table_open": "<table>", "table_close": "</table>", "row_open":"<tr>"}
+
 PDF_HEADERS = {
     "title": "h1",
     "sectionHeading": "h2"
 }
 
+class TokenEstimator(object):
+    GPT2_TOKENIZER = tiktoken.get_encoding("gpt2")
+    CHATGPT_TOKENIZER = tiktoken.get_encoding("cl100k_base")
+
+    def estimate_tokens(self, text: Union[str, List]) -> int:
+        if isinstance(text, str):
+            # print(f"The length is {len(self.GPT2_TOKENIZER.encode(text, allowed_special='all'))}")
+            return len(self.GPT2_TOKENIZER.encode(text, allowed_special="all"))
+        else:
+            # https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb
+            tokens_per_message = 4  # every message follows <|start|>{role/name}\n{content}<|end|>\n
+            tokens_per_name = -1    # if there's a name, the role is omitted
+            num_tokens = 0
+            for message in text:
+                num_tokens += tokens_per_message
+                for key, value in message.items():
+                    num_tokens += len(self.CHATGPT_TOKENIZER.encode(value))
+                    if key == "name":
+                        num_tokens += tokens_per_name
+            num_tokens += 3  # every reply is primed with <|start|>assistant<|message|>
+            return num_tokens
+
+    def construct_tokens_with_size(self, tokens: str, numofTokens: int) -> str:
+        newTokens = self.GPT2_TOKENIZER.decode(
+            self.GPT2_TOKENIZER.encode(tokens, allowed_special="all")[:numofTokens]
+        )
+        return newTokens
+    
+TOKEN_ESTIMATOR = TokenEstimator()
+
+
+class PdfTextSplitter(TextSplitter):
+    def __init__(self, length_function: Callable[[str], int] =TOKEN_ESTIMATOR.estimate_tokens, separator: str = "\n\n", **kwargs: Any):
+        """Create a new TextSplitter for htmls from extracted pdfs."""
+        super().__init__(**kwargs)
+        self._table_tags = HTML_TABLE_TAGS
+        self._separators = separator or ["\n\n", "\n", " ", ""]
+        self._length_function = length_function
+
+    def extract_caption(self, s, type):

```

Comment:
Just for PMs - any latency number per 1024 tokens? Just estimate

Reply:
todo